src/devices/bus/megadrive/paprium.cpp

```diff
@@ -0,0 +1,946 @@
+	save_item(NAME(ppm_unk_data2_addr));
+
+	//	save_item(NAME(ppm_bgm_tracks_offsets));
+	save_item(NAME(ppm_bgm_tracks_base_addr));
+	save_item(NAME(ppm_bgm_unpack_addr));
+
+	//	save_item(NAME(ppm_sfx_data));
+	save_item(NAME(ppm_sfx_base_addr));
+
+	//	save_item(NAME(ppm_gfx_blocks_offsets));
+	save_item(NAME(ppm_gfx_blocks_base_addr));
+	save_item(NAME(ppm_max_gfx_block));
+
+	save_item(NAME(ppm_audio_bgm_volume));
+	save_item(NAME(ppm_audio_sfx_volume));
+	save_item(NAME(ppm_audio_config));
+}
+
+void md_rom_paprium_device::device_reset()
+{
+	// currently loading MAX10 data over rom area then copies it to ram block,
+	// this should be changed later on to add proper dataarea for this to load into
+
+	// did we decode yet?
+	if (m_rom[0x8000 / 2])
+	{
+		uint16_t key1 = m_rom[0x8000 / 2];  // 0x0000 reset stack ptr
+		uint16_t key2 = m_rom[0xbd000 / 2]; // blank area
+
+		// top 8K overlayed with MAX10 dual port ram module
+		// for (uint32_t addr = 0; addr < 0x2000 / 2; addr++)
+		//	m_rom[addr] ^= (key2 | bitswap<16>(addr & 0xff, 15, 1, 14, 6, 13, 2, 12, 0, 11, 3, 10, 4, 9, 7, 8, 5));
+		for (uint32_t addr = 0x2000 / 2; addr < 0x10000 / 2; addr++)
+			m_rom[addr] ^= (key1 | bitswap<16>(addr & 0xff, 15, 1, 14, 6, 13, 2, 12, 0, 11, 3, 10, 4, 9, 7, 8, 5));
+		for (uint32_t addr = 0x10000 / 2; addr < 0x800000 / 2; addr++)
+			m_rom[addr] ^= (key2 | bitswap<16>(addr & 0xff, 15, 1, 14, 6, 13, 2, 12, 0, 11, 3, 10, 4, 9, 7, 8, 5));
+
+		// copy data to dual port ram
+		for (uint16_t x = 0; x < 0x2000 / 2; x++)
+			ppm_dual_port_ram[x] = m_rom[x];
+
+		// check ROM version & patch a few things
+		// (hold B on startup to display version ingame)
+		switch (m_rom[0x1000a / 2])
+		{
+		case 0x2e7f:
+			// boot check
+			ppm_dual_port_ram[0x1d1c / 2] = 0x0004;	 // 0xc00004
+			ppm_dual_port_ram[0x1d2c / 2] |= 0x0100; // switch to beq.s (0x670e)
+
+			// post splash jump location
+			ppm_dual_port_ram[0x1562 / 2] = 0x0001;
+			ppm_dual_port_ram[0x1564 / 2] = 0x0100;
+
+			// emulator check
+			m_rom[0x81104 / 2] = 0x4e71;
+
+			// mame pretends to be version 1 megadrive but does not implement
+			// proper TMSS register & rom, this will fix the issue being detected
+			m_rom[0xbca5e / 2] = 0x6006; // bra.s to the correct location
+			break;
+		default:
+			printf("UNKNOWN PAPRIUM VERSION 0x%04X\n", m_rom[0x1000a / 2]);
+			break;
+		}
+	}
+
+	// clear/setup registers
+	ppm_interface->reg_command = 0;
+	ppm_interface->reg_status_1 = 0;
+	ppm_interface->reg_status_2.word = 0;
+	ppm_interface->reg_status_2.bits.megawire_status = 7; // let's pretend MW is plugged & connected
+
+	ppm_sdram_pointer = ppm_sdram;
+	ppm_sdram_window_enabled = false;
+
+	ppm_vram_max_slot = 0;
+}
+
+uint16_t md_rom_paprium_device::read(offs_t offset)
+{
+	if (offset < sizeof(ppm_dual_port_ram) / 2)
+		return ppm_dual_port_ram[offset];
+
+	if ((offset >= 0xc000 / 2) && (offset < 0x10000 / 2) && ppm_sdram_window_enabled)
+		return *ppm_sdram_pointer++;
```

Comment:
The increment needs to guard against `!machine().side_effects_disabled()`

Without it, havoc will happen with MAME debugger open + CTRL+M viewer at 0x00c000 onward.